Makefile

```diff
@@ -0,0 +1,80 @@
+PREFIX ?= /usr/local
+
+.PHONY: release debug profile all clean install
+OPTLEVEL ?= -O3
+
+LIBS += sdl2 libpng libjpeg gl glew openal mad
+CXXFLAGS += -std=c++11 -Wall
+CXXFLAGS += $(shell pkg-config $(LIBS) --cflags)
+LDFLAGS += $(shell pkg-config $(LIBS) --libs) -lpthread
+
+PROG = endless-sky
+PROG_DEBUG = $(PROG)-debug
+PROG_PROF = $(PROG)-profile
+DIR_SRC = source
+DIR_BUILD = build
+DIR_RELEASE = $(DIR_BUILD)/release
+DIR_DEBUG = $(DIR_BUILD)/debug
+DIR_PROF = $(DIR_BUILD)/profile
+
+SRCS = $(wildcard $(DIR_SRC)/*.cpp)
+OBJS = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_RELEASE)/%.o,$(SRCS))
+DEPS = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_RELEASE)/%.d,$(SRCS))
+OBJS_DEBUG = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_DEBUG)/%.o,$(SRCS))
+DEPS_DEBUG = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_DEBUG)/%.d,$(SRCS))
+OBJS_PROF = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_PROF)/%.o,$(SRCS))
+DEPS_PROF = $(patsubst $(DIR_SRC)/%.cpp,$(DIR_PROF)/%.d,$(SRCS))
+
+$(info mkdir -p $(DIR_RELEASE) $(DIR_DEBUG) $(DIR_PROF))
+$(shell mkdir -p $(DIR_RELEASE) $(DIR_DEBUG) $(DIR_PROF))
+
+$(DIR_RELEASE)/%.o: $(DIR_SRC)/%.cpp
+	$(CXX) $(OPTLEVEL) $(CXXFLAGS) -o $@ -c -MMD $<
+
+release: $(OBJS)
```

Comment:
Currently this will cause `make` to link the game unconditionally on every invocation (i.e. the build is never clean). Because the binary for all build modes is the same, there is the problem that when building A, then B, then A again the build system would need to know to relink the game in mode A even though all timestamps are up to date.

I'm not sure there is a clean way to do this in Make. Scons seems to be aware of this and relinks the binary as needed, while still allowing clean rebuilds. I guess you could link the binary in the build directory, then just copy it over in the top-level target, which would save the effort of linking.